qualtran/bloqs/mod_arithmetic/mod_multiplication_test.py

```diff
@@ -0,0 +1,126 @@
+
+
+@pytest.mark.parametrize('dtype', [QUInt, QMontgomeryUInt])
+@pytest.mark.parametrize(
+    ['prime', 'bitsize'], [(p, n) for p in (13, 17, 23) for n in range(p.bit_length(), 10)]
+)
+def test_moddbl_decomposition(dtype, bitsize, prime):
+    b = ModDbl(dtype(bitsize), prime)
+    qlt_testing.assert_valid_bloq_decomposition(b)
+
+
+@pytest.mark.parametrize('dtype', [QUInt, QMontgomeryUInt])
+@pytest.mark.parametrize(
+    ['prime', 'bitsize'], [(p, n) for p in (13, 17, 23) for n in range(p.bit_length(), 10)]
+)
+def test_moddbl_bloq_counts(dtype, bitsize, prime):
+    b = ModDbl(dtype(bitsize), prime)
+    qlt_testing.assert_equivalent_bloq_counts(b, [ignore_alloc_free, ignore_split_join])
+
+
+def test_moddbl_cost():
+    n, p = sympy.symbols('n p')
+    b = ModDbl(QMontgomeryUInt(n), p)
+    cost = get_cost_value(b, QECGatesCost()).total_t_and_ccz_count()
+
+    # Litinski 2023 https://arxiv.org/abs/2306.08585
+    # Figure/Table 8. Lists modular doubling as 2n toffoli.
+    assert cost['n_t'] == 0
+    assert cost['n_ccz'] == 2 * n + 1
+
+
+@pytest.mark.slow
+@pytest.mark.parametrize('dtype', [QUInt, QMontgomeryUInt])
+@pytest.mark.parametrize(
+    ['prime', 'bitsize', 'k'],
+    [(p, n, k) for p in (13, 17, 23) for n in range(p.bit_length(), 10) for k in range(1, p)],
+)
+def test_cmodmulk_classical_action(dtype, bitsize, prime, k):
+    b = CModMulK(dtype(bitsize), k=k, mod=prime)
+    qlt_testing.assert_consistent_classical_action(b, ctrl=(0, 1), x=range(prime))
+
+
+@pytest.mark.parametrize('k', range(1, 13))
+def test_cmodmulk_classical_action_fast(k):
+    b = CModMulK(QMontgomeryUInt(4), k=k, mod=13)
+    qlt_testing.assert_consistent_classical_action(b, ctrl=(0, 1), x=range(13))
+
+
+@pytest.mark.parametrize('dtype', [QUInt, QMontgomeryUInt])
+@pytest.mark.parametrize(
+    ['prime', 'bitsize', 'k'],
+    [(p, n, k) for p in (13, 17, 23) for n in range(p.bit_length(), 10) for k in range(1, p)],
+)
+def test_cmodmulk_decomposition(dtype, bitsize, prime, k):
+    b = CModMulK(dtype(bitsize), k, prime)
+    qlt_testing.assert_valid_bloq_decomposition(b)
+
+
+@pytest.mark.parametrize('dtype', [QUInt, QMontgomeryUInt])
+@pytest.mark.parametrize(
+    ['prime', 'bitsize', 'k'],
+    [(p, n, k) for p in (13, 17, 23) for n in range(p.bit_length(), 10) for k in range(1, p)],
+)
+def test_cmodmulk_bloq_counts(dtype, bitsize, prime, k):
+    b = CModMulK(dtype(bitsize), k, prime)
+    ssa = SympySymbolAllocator()
+    my_k = ssa.new_symbol('k')
+
+    def generalizer(bloq):
+        if isinstance(bloq, CtrlScaleModAdd):
+            return attrs.evolve(bloq, k=my_k)
+        return bloq
+
+    qlt_testing.assert_equivalent_bloq_counts(
+        b, [ignore_alloc_free, ignore_split_join, generalizer]
+    )
+
+
+@pytest.mark.parametrize('example', [_moddbl_small, _moddbl_large, _modmul_symb, _modmul])
+def test_examples(bloq_autotester, example):
+    bloq_autotester(example)
```

Comment:
haha this is a bit much. At least split up the moddbl and modmul examples into separate unit tests. Tests should be DAMP not DRY! 

https://testing.googleblog.com/2019/12/testing-on-toilet-tests-too-dry-make.html

specifically, it's useful to have distinct unit tests to narrow in on failures should they arise

Reply:
bump

Reply:
done